lib/Conversion/TritonGPUToLLVM/Utility.cpp

```diff
@@ -907,4 +907,137 @@ Value mxfpScaleBf16(RewriterBase &rewriter, Location loc, Value v,
 };
 
 } // namespace LLVM
+
+SharedMemoryObject
+getExpandedSharedMemoryObject(ConversionPatternRewriter &rewriter, Location loc,
+                              SharedMemoryObject smemObj,
+                              ArrayRef<int64_t> shape) {
+  auto strides = smemObj.getStrides();
+  auto offsets = smemObj.getOffsets();
+  auto rank = strides.size();
+  if (rank == 3)
+    return smemObj;
+  auto expandedStrides = insertValue(strides, 0, i32_val(shape[0] * shape[1]));
+  auto expandedOffsets = insertValue(offsets, 0, i32_val(0));
+  auto expandedSmemObj =
+      SharedMemoryObject(smemObj.getBase(), smemObj.getBaseElemType(),
+                         expandedStrides, expandedOffsets);
+  return expandedSmemObj;
+}
+
+CTALayoutAttr getExpandedCTALayout(MLIRContext *ctx,
+                                   CTALayoutAttr ctaLayoutAttr) {
+  auto rank = ctaLayoutAttr.getCTAsPerCGA().size();
+  if (rank == 3) {
+    return ctaLayoutAttr;
+  }
+
+  auto ctasPerCGA3d =
+      insertValue<unsigned>(ctaLayoutAttr.getCTAsPerCGA(), rank, 1);
+  auto ctasSplitNum3d =
+      insertValue<unsigned>(ctaLayoutAttr.getCTASplitNum(), rank, 1);
+  auto ctaOrder3d =
+      insertValue<unsigned>(ctaLayoutAttr.getCTAOrder(), rank, rank);
+  auto expandedCTALayout = triton::gpu::CTALayoutAttr::get(
+      ctx, ctasPerCGA3d, ctasSplitNum3d, ctaOrder3d);
+  return expandedCTALayout;
+}
+
+/// Expand layout of dot operands and dot results to 3d variant.
+///
+/// If given layout describes 3d tensor, return it without change.
+/// If given layout describes 2d tensor, create new layout
+/// describing 3d tensor by adding batch = 1.
+Attribute getExpandedEncoding(Attribute encoding) {
+  auto ctx = encoding.getContext();
+  if (auto sharedEncoding = mlir::dyn_cast<SharedEncodingAttr>(encoding)) {
+    auto order = sharedEncoding.getOrder();
+    auto rank = order.size();
+    if (rank == 3) {
+      return sharedEncoding;
+    }
```

Comment:
This early return is the same for all layouts. Can you move it out of this if-else block?

Reply:
It's not identical for all branches and it is not needed for the last branch, so I left it as it.

Reply:
There is this `getOrder` [API](https://github.com/triton-lang/triton/blob/24b8d431851ed0ae77b7addfebf83c08ec316514/include/triton/Dialect/TritonGPU/IR/Dialect.h#L83) that works for all layouts. You can query the order of the layout, and if the size of the order is 3, early return.